src/core/Akka/Util/Internal/AtomicState.cs

```diff
@@ -73,149 +74,78 @@ await Task
         /// <summary>
         /// Shared implementation of call across all states.  Thrown exception or execution of the call beyond the allowed
         /// call timeout is counted as a failed call, otherwise a successful call
-        /// 
-        /// NOTE: In .Net there is no way to cancel an uncancellable task. We are merely cancelling the wait and marking this
-        /// as a failure.
-        /// 
-        /// see http://blogs.msdn.com/b/pfxteam/archive/2011/11/10/10235834.aspx 
         /// </summary>
-        /// <typeparam name="T">TBD</typeparam>
-        /// <param name="task">Implementation of the call</param>
-        /// <returns>result of the call</returns>
+        /// <param name="task"><see cref="Task"/> Implementation of the call</param>
+        /// <returns><see cref="Task"/> containing the result of the call</returns>
         public async Task<T> CallThrough<T>(Func<Task<T>> task)
         {
-            var deadline = DateTime.UtcNow.Add(_callTimeout);
-            ExceptionDispatchInfo capturedException = null;
-            T result = default(T);
+            var result = default(T);
             try
             {
-                result = await task().ConfigureAwait(false);
+                result = await task().WaitAsync(_callTimeout).ConfigureAwait(false);
```

Comment:
Got it - this fix makes sense to me. Can't trust the `task()` to necessarily cancel itself. Have to control whether or not it completes on-time via our own abstractions.

Reply:
Yep, this is where we were potentially awaiting a task indefinitely (depending on how long the task would take) despite the `CallTimeout`